<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [375]:
import pandas as pd
import psycopg2

In [376]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [377]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [378]:
# текст запроса
query_3_1 = f'''SELECT count(id) as "count total vacancies"
            FROM vacancies
            '''

In [379]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count total vacancies
0                  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [380]:
# текст запроса
query_3_2 = f'''SELECT count(id) as "count total employers"
            FROM employers
            '''

In [381]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count total employers
0                  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [382]:
# текст запроса
query_3_3 = f'''SELECT count(id) as "count of regions"
            FROM areas
            '''

In [383]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count of regions
0              1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [384]:
# текст запроса
query_3_4 = '''SELECT COUNT(DISTINCT industry_id) as "count of industries"
			FROM employers_industries
			'''

In [385]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count of industries
0                  294

***

In [386]:
# выводы по предварительному анализу данных
queries = [query_3_1, query_3_2, query_3_3, query_3_4]
results = []

for q in queries:
    results.append(pd.read_sql_query(q, connection))

print("\n Общий вывод на все 4 вопроса приставлен ниже в виде таблицы:")
summary_df = pd.concat(results, axis=1)
summary_df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\250783729.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  results.append(pd.read_sql_query(q, connection))



 Общий вывод на все 4 вопроса приставлен ниже в виде таблицы:


count total vacancies  count total employers  count of regions  \
0                  49197                  23501              1362   

   count of industries  
0                  294

База данных HeadHunters обладает большим количством работодателей (23501). Количество открытых вакансий 49197, в 294-х различных сферах деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [387]:
# текст запроса
query_4_1 = '''SELECT areas.name as "area name",
                COUNT(vacancies.id) as "count of vacancies"
			FROM vacancies
            JOIN areas ON vacancies.area_id = areas.id
            GROUP BY areas.name
            ORDER BY COUNT(vacancies.id) DESC
			limit 5
            '''

In [388]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area name  count of vacancies
0           Москва                5333
1  Санкт-Петербург                2851
2            Минск                2112
3      Новосибирск                2006
4           Алматы                1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [389]:
# текст запроса
query_4_2 = '''SELECT COUNT(id) as "vacancies with salary"
			FROM vacancies
            WHERE salary_from IS NOT null  OR salary_to IS NOT null 
            '''

In [390]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


vacancies with salary
0                  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [391]:
# текст запроса
query_4_3 = '''SELECT ROUND(AVG(salary_from)) AS "avg salary_from",
            ROUND(AVG(salary_to)) AS "avg salary_to"
            FROM vacancies
            WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
            '''

In [392]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg salary_from  avg salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [393]:
# текст запроса
query_4_4 = '''SELECT
            schedule,
            employment,
            COUNT(id) AS "vacancy_count"
            FROM vacancies
            GROUP BY schedule, employment
            ORDER BY vacancy_count DESC
            LIMIT 2
            '''

In [394]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment  vacancy_count
0       Полный день  Полная занятость          35367
1  Удаленная работа  Полная занятость           7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [395]:
# текст запроса
query_4_5 = '''SELECT experience,
            COUNT(id) AS vacancy_count
            FROM vacancies
            GROUP BY experience
            ORDER BY vacancy_count ASC
            '''

In [396]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancy_count
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

In [397]:
# выводы по детальному анализу вакансий

Выводы:
1. Как и ожидалось наибольшее количество вакансий наблюдается в крупных городах (Москва, Питер и тд)
2. Только половина вакансий имеют ЗП указанной в заявке
3. Границы средней ЗП 71-110 тыс руб
4. Большинство заявок на полный день и полную занятость
5. Самый часто запрашиваемый опыт работы 1-3 года

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [398]:
# текст запроса
query_5_1 = '''SELECT employers.name as "employer name",
                COUNT(vacancies.id) as "count of vacancies"
			FROM vacancies
            JOIN employers ON vacancies.employer_id = employers.id
            GROUP BY employers.name
            ORDER BY COUNT(vacancies.id) DESC
			limit 5
            '''

In [399]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


employer name  count of vacancies
0         Яндекс                1933
1     Ростелеком                 491
2       Тинькофф                 444
3           СБЕР                 428
4  Газпром нефть                 331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [400]:
# текст запроса
query_5_2 = '''SELECT areas.name AS "region",
			COUNT(DISTINCT vacancies.employer_id) AS "employers_count",
			COUNT(vacancies.id) AS "vacancies_count"
			FROM areas
			LEFT JOIN vacancies ON areas.id = vacancies.area_id
			GROUP BY areas.name
			ORDER BY vacancies_count DESC, employers_count DESC
			'''

In [401]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region  employers_count  vacancies_count
0                    Москва             2320             5333
1           Санкт-Петербург             1451             2851
2                     Минск              851             2112
3               Новосибирск              802             2006
4                    Алматы              798             1892
...                     ...              ...              ...
1357             Нефтекумск                0                0
1358          Краснозёрское                0                0
1359                    Нея                0                0
1360          Краснозерское                0                0
1361  Нижегородская область                0                0

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [402]:
# текст запроса
query_5_3 = '''SELECT
            employers.name AS "employer_name",
            COUNT(DISTINCT vacancies.area_id) AS "regions_count"
            FROM vacancies
            JOIN employers ON vacancies.employer_id = employers.id
            GROUP BY employers.name
            ORDER BY regions_count DESC
            LIMIT 10
            '''

In [403]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer_name  regions_count
0                      Яндекс            181
1                  Ростелеком            152
2                  Спецремонт            116
3      Поляков Денис Иванович             88
4                   ООО ЕФИН              71
5                  Совкомбанк             63
6                         МТС             55
7  ЭФКО, Управляющая компания             49
8                Почта России             48
9                        КРОН             48

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [404]:
# текст запроса
query_5_4 = '''SELECT COUNT(employers.id) AS employers_without_industry
            FROM employers
            LEFT JOIN employers_industries ON employers.id = employers_industries.employer_id
            WHERE employers_industries.industry_id IS NULL
            '''

In [405]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employers_without_industry
0                        8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [406]:
# текст запроса
query_5_5 = '''SELECT employers.name AS "company_name"
            FROM employers
            JOIN employers_industries ON employers.id = employers_industries.employer_id
            GROUP BY employers.id, employers.name
            HAVING COUNT(DISTINCT employers_industries.industry_id) = 4
            ORDER BY employers.name
            OFFSET 2 LIMIT 1
            '''

In [407]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


company_name
0         2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [408]:
# текст запроса
query_5_6 = '''SELECT COUNT(DISTINCT employers.id) AS "Разработка программного обеспечения"
            FROM employers
            JOIN employers_industries ON employers.id = employers_industries.employer_id
            JOIN industries ON employers_industries.industry_id = industries.id
            WHERE industries.name = 'Разработка программного обеспечения'
            '''

In [409]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


Разработка программного обеспече
0                              3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [428]:
# код для получения списка городов-милионников
import requests
import pandas as pd

# Загружаем HTML-страницу
url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"
response = requests.get(url)
tables = pd.read_html(response.text)

# Находим таблицу с городами-миллионниками (первая таблица на странице)
million_cities_table = tables[0]

# Извлекаем названия городов из первого столбца, пропуская заголовок
million_cities = million_cities_table.iloc[:, 1].tolist()

# Убираем сноски и лишние пробелы
million_cities = [city.split('[')[0].strip() for city in million_cities]

print(million_cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3469690117.py:8: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(response.text)


In [463]:
# текст запроса
query_5_7 = f"""
SELECT
areas.name AS "region",
COUNT(vacancies.id) AS "vacancies_count"
FROM vacancies
JOIN employers ON vacancies.employer_id = employers.id
JOIN areas ON vacancies.area_id = areas.id
WHERE employers.name = 'Яндекс'
  AND areas.name IN ({', '.join(["'%s'" % city for city in million_cities])})
GROUP BY areas.name
ORDER BY vacancies_count ASC
"""


In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)

# добавляем строку Total
total_vacancies = df['vacancies_count'].sum()
df.loc[len(df)] = ['Total', total_vacancies]
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\1477916998.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


region  vacancies_count
0              Омск               21
1         Челябинск               22
2        Красноярск               23
3         Волгоград               24
4             Пермь               25
5            Казань               25
6    Ростов-на-Дону               25
7               Уфа               26
8            Самара               26
9         Краснодар               30
10          Воронеж               32
11      Новосибирск               35
12  Нижний Новгород               36
13     Екатеринбург               39
14  Санкт-Петербург               42
15           Москва               54
16            Total              485

***

In [498]:
# выводы по анализу работодателей

- Наибольшее количество вакансий в "Яндекс". Топ 3 компании: Яндекс, Ростелеком, Тинькофф
- Большинство крупных компаний нанимают одновременно в нескольких регионах 
- Многие компании работают в различных сферах деятельности (более одной)
- 3533 работодателей ищут сотрудников в сфере IT
- Был разобран пример "Яндекс" по количеству вакансий в городах с населением более 1 млн жителей. Общее число вакансий в этих городах 485.
- Рынок труда заинтересован в сотрудниках сферы IT


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [514]:
# текст запроса
query_6_1 = """
SELECT COUNT(*) AS "vacancies_with_data"
FROM vacancies
WHERE LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%'
"""

In [530]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


vacancies_with_data
0                 1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [559]:
# текст запроса
query_6_2 = """
SELECT COUNT(*) AS "junior_ds_vacancies"
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%'
    OR LOWER(name) LIKE '%data science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%'
)
AND (
    LOWER(name) LIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
)
"""

In [560]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


junior_ds_vacancies
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [582]:
# текст запроса
query_6_3 = """
SELECT COUNT(*) AS "SQL and postgres skills"
FROM vacancies
WHERE (
	LOWER(name) LIKE '%data scientist%'
	OR LOWER(name) LIKE '%data science%'
	OR LOWER(name) LIKE '%исследователь данных%'
	OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
	OR LOWER(name) LIKE '%machine learning%'
	OR LOWER(name) LIKE '%машинн%обучен%'
)
AND (
	LOWER(key_skills) LIKE '%sql%'
	OR LOWER(key_skills) LIKE '%postgres%'
)
"""


In [583]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


SQL and postgres skills
0                      229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [701]:
# текст запроса
query_6_4 = """
SELECT COUNT(*) AS "SQL and postgres skills"
FROM vacancies
WHERE (
	LOWER(name) LIKE '%data scientist%'
	OR LOWER(name) LIKE '%data science%'
	OR LOWER(name) LIKE '%исследователь данных%'
	OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
	OR LOWER(name) LIKE '%machine learning%'
	OR LOWER(name) LIKE '%машинн%обучен%'
)
AND (
	LOWER(key_skills) LIKE '%python%'
)
"""

In [702]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


SQL and postgres skills
0                      357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [732]:
# текст запроса
query_6_5 = """
SELECT 
ROUND(AVG(CHAR_LENGTH(key_skills) - CHAR_LENGTH(replace(key_skills, CHR(9), '')) + 1), 2) AS "Количество Слов Навыков"

FROM vacancies
WHERE (
    lower(name) LIKE '%data scientist%' 
    OR lower(name) LIKE '%data science%' 
    OR lower(name) LIKE '%исследователь данных%' 
    OR (name LIKE '%ML%' AND lower(name) NOT LIKE'%html%') 
    OR lower(name) LIKE '%machine learning%' 
    OR lower(name) LIKE '%машинн%обучен%'
) 
AND (key_skills IS NOT NULL AND key_skills != '')
"""

In [733]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df
    

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\1850607359.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Количество Слов Навыков
0                     6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [793]:
# текст запроса
query_6_6 = """
SELECT
    experience,
    ROUND(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0)
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%'
    OR LOWER(name) LIKE '%data science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%'
    AND (key_skills IS NOT NULL)
)
GROUP BY experience
ORDER BY 1
"""

In [794]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  145236.0
3       От 3 до 6 лет  243115.0

***

In [796]:
# выводы по предметному анализу

- 1771 вакансий связанных с работой с данными
- 51 вакансий на Junior Data Scientist (без опыта работы или стажировка)
- 229 вакансий, указывают знание SQL & postgres в графе умений и навыков
- 357 случаях указано знание языка Python 
- В среднем ожидается 6 и более навыков от соискателей
- Был проанализирован оклад труда для Data Scientist в зависимости от опыта работы. Junior порядка 74т.р., с последующим повышение по мере получения опыта

# Общий вывод по проекту

In [797]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

#### Общий вывод на основе проделанных исследований
1. База данных HeadHunter содержит значительный объем информации: почти 50 тысяч вакансий, более 23 тысяч работодателей, охватывающих 1362 региона и 294 сферы деятельности.
2. Наибольшее количество вакансий сосредоточено в крупных городах, таких как Москва и Санкт-Петербург.
3. Только около половины вакансий содержат информацию о зарплате, средние значения которой варьируются от 71 до 110 тысяч рублей.
4. Большинство вакансий предполагают полный рабочий день и полную занятость, а наиболее востребованный опыт — 1-3 года.
5. Крупные работодатели, такие как Яндекс, публикуют вакансии сразу в нескольких регионах и сферах деятельности.
6. В сфере Data Science наблюдается высокий спрос на специалистов, особенно с навыками Python, SQL и машинного обучения. Для Junior Data Scientist открыто порядка 50 вакансий.
7. В среднем в вакансиях для DS указывается 6 и более ключевых навыков, а уровень зарплаты растет с увеличением опыта.

#### Возможные направления для дальнейших исследований:
- Анализ динамики изменения числа вакансий и зарплат по годам
- Сравнение востребованности различных профессий и навыков в разных регионах.
- Прогнозирование спроса на специалистов в сфере IT и Data Science на основе текущих трендов.
- Исследование карьерных траекторий: как часто работодатели ищут специалистов с опытом работы в смежных областях.
- Кластеризация вакансий по ключевым навыкам для выявления новых профессиональных направлений

In [800]:
# Дополнительный анализ: распределение вакансий по регионам и сферам деятельности

# 1. Топ-10 регионов по количеству вакансий
query_top_regions = '''
SELECT areas.name AS "region", COUNT(vacancies.id) AS "vacancies_count"
FROM vacancies
JOIN areas ON vacancies.area_id = areas.id
GROUP BY areas.name
ORDER BY vacancies_count DESC
LIMIT 10
'''
top_regions_df = pd.read_sql_query(query_top_regions, connection)
print("Топ-10 регионов по количеству вакансий:")
display(top_regions_df)

# 2. Топ-10 сфер деятельности по количеству вакансий
query_top_industries = '''
SELECT industries.name AS "industry", COUNT(vacancies.id) AS "vacancies_count"
FROM vacancies
JOIN employers ON vacancies.employer_id = employers.id
JOIN employers_industries ON employers.id = employers_industries.employer_id
JOIN industries ON employers_industries.industry_id = industries.id
GROUP BY industries.name
ORDER BY vacancies_count DESC
LIMIT 10
'''
top_industries_df = pd.read_sql_query(query_top_industries, connection)
print("Топ-10 сфер деятельности по количеству вакансий:")
display(top_industries_df)

# 3. Распределение вакансий по требуемому опыту
query_experience_dist = '''
SELECT experience, COUNT(id) AS "vacancies_count"
FROM vacancies
GROUP BY experience
ORDER BY vacancies_count DESC
'''
experience_dist_df = pd.read_sql_query(query_experience_dist, connection)
print("Распределение вакансий по требуемому опыту:")
display(experience_dist_df)

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3414301022.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_regions_df = pd.read_sql_query(query_top_regions, connection)


Топ-10 регионов по количеству вакансий:


region  vacancies_count
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892
5     Екатеринбург             1698
6  Нижний Новгород             1670
7           Казань             1415
8        Краснодар             1301
9           Самара             1144

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3414301022.py:27: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_industries_df = pd.read_sql_query(query_top_industries, connection)


Топ-10 сфер деятельности по количеству вакансий:


industry  vacancies_count
0                Разработка программного обеспечения            12499
1  Системная интеграция,  автоматизации технологи...            11034
2  Интернет-компания (поисковики, платежные систе...             6413
3                                               Банк             2742
4                                    Мобильная связь             1585
5                                 Кадровые агентства             1143
6                              Консалтинговые услуги             1137
7                                 Интернет-провайдер             1095
8                       Розничная сеть (продуктовая)             1034
9                                Фиксированная связь             1028

C:\Users\tgorbunov\AppData\Local\Temp\ipykernel_22464\3414301022.py:38: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  experience_dist_df = pd.read_sql_query(query_experience_dist, connection)


Распределение вакансий по требуемому опыту:


experience  vacancies_count
0  От 1 года до 3 лет            26152
1       От 3 до 6 лет            14511
2           Нет опыта             7197
3         Более 6 лет             1337